### PR TITLE
network tests: wire the local mesh by direct dialing instead of mDNS

### PR DESCRIPTION
**Pull request by iurii-ssv**
## Problem

`network/p2p` and `network/topics` formed their local test networks through mDNS. Wherever multicast is blocked for the test process (sandboxed shells, macOS local-network privacy, some CI runners), the nodes never found each other, `CreateAndStartLocalNet` gave up after its three attempts, and closing the mDNS service wedged on the way out, which is what the timed-out-close workarounds in both suites were papering over.

## Change

- The test harnesses now dial their nodes into a full mesh by address, so the mesh forms the same way on every machine. `connectMesh` dials every edge exactly once and lets no node take more than half of its edges inbound, which is what the connection gater's inbound limit allows.
- The local net runs no discovery: a new `none` discovery mode installs `discovery.Disabled`, a `Service` that discovers nothing, so a node keeps only its trusted, inbound and directly connected peers. The mode is documented in the `Discovery` option's description.
- The external-IP lookup, and the explicit IP listener it feeds, now happen only for `discv5`, the mode that advertises it. A node running `mdns` or `none` no longer needs an external IP at all, so it cannot fail setup because its interfaces could not be enumerated.
- The per-process mDNS service tag (`MdnsDiscoveryTag`, the `serviceTag` parameter of `NewLocalDiscovery`) only existed to keep concurrent test runs from cross-discovering each other; with no test using mDNS it is gone, and the timed-out-close workarounds with it.

`mdns` itself is untouched and remains the local docker setup's discovery (see `docs/DEV_GUIDE.md`).

## Verification

In a shell where a multicast loopback probe receives nothing (the state that failed before):

| suite | before | after |
| --- | --- | --- |
| `network/p2p` | `TestP2pNetwork_MessageValidation`: "network didn't start after 3 attempts", 155s | ok, 9.5s |
| `network/topics` | `TestTopicManager/happy_flow`: "peers never reached connection threshold 3", 38s | ok, 9.7s |
| `network/discovery` | ok | ok |

`TestP2pNetwork_SubscribeBroadcast`, `TestP2pNetwork_MessageValidation` and `TestTopicManager` pass on three consecutive runs each; the whole `network/...` tree is green; `golangci-lint` and `misspell` report nothing on the touched packages.

`discovery.Disabled` has a unit test of its own in `network/discovery`.

## Note for reviewers

`none` is a small piece of new operator-facing config surface. The alternative was a code-only test hook on `Config`, which is the pattern `MdnsDiscoveryTag` followed; a named mode seemed the more honest way to express "this node runs no discovery", and it is a legitimate setup for a node that only ever talks to its `TrustedPeers`. Happy to switch to the hook if the extra surface is unwanted.

### CODE REVIEW

**Comment by iurii-ssv** [comment]
@greptile pls re-review

File: network/p2p/testutils.go [hunk 1]
```diff
@@ -129,28 +117,27 @@ func CreateAndStartLocalNet(pCtx context.Context, logger *zap.Logger, options Lo
...
-// indefinitely on a multicast socket write — so once the timeout elapses we stop
-// waiting and let the next attempt proceed with fresh hosts and ports; an
-// abandoned Close is reclaimed when the test process exits.
-func closeNodesWithTimeout(logger *zap.Logger, nodes []network.P2PNetwork, timeout time.Duration) {
-	var eg errgroup.Group
-	for _, node := range nodes {
-		eg.Go(func() error { return node.Close() })
+// connectMesh dials the nodes into a full mesh by address, so the local network forms without discovery.
+// Node i dials the nodes up to half a ring ahead of it: every edge is dialed exactly once, and no node takes
+// more than half of its edges inbound, which is what the connection gater's inbound limit allows.
```
Comment (momosh-ssv) [inline network/p2p/testutils.go:122]: Worth considering: this holds up to n=17 and breaks at n=18. Max inbound per node is floor(n/2), every peer dials from 127.0.0.1, and the gater's per-IP leaky bucket is `ipLimitBurst = 8` (`network/peers/connections/conn_gater.go`), so the 9th inbound edge gets refused and `connectMesh` fails all three attempts. I ran it: 5, 8, 13, 14, 16 and 17 form a full mesh on the first attempt, 18 fails deterministically with `all dials failed ... connection reset by peer`.

There is also a hidden coupling behind "which is what the connection gater's inbound limit allows": `inboundLimit()` is `int(MaxPeers*0.5)` and `NewNetConfig` sets `MaxPeers = options.Nodes`, so the mesh's max inbound equals the limit exactly, with zero headroom. A future caller that sets `MaxPeers` independently breaks the mesh with a misleading dial error.

`DisableIPRateLimit: true` in `NewNetConfig` removes both (it short-circuits `InterceptAccept`).

The same `MaxPeers == Nodes` coupling also means `peersTrimming` trims without a floor once it first fires at t=30s: the skip guard `len(connectedPeers) <= MaxPeers-maxPeersToDrop` is `3 <= 0`, so it runs with `trimInboundOnly = false`, and trimmed peers are then blocked for 30 min by `trimmedRecently`. Nothing re-dials now that mDNS is gone, so a test running past 30s would degrade with no recovery. Current tests finish first, so this is latent rather than broken.

File: network/p2p/p2p_setup.go [hunk 2]
```diff
@@ -298,14 +298,18 @@ func (n *p2pNetwork) setupDiscovery() error {
...
-		disc, err = discovery.NewLocalDiscovery(n.ctx, logger, n.Host(), n.cfg.MdnsDiscoveryTag)
+		disc, err = discovery.NewLocalDiscovery(n.ctx, logger, n.Host())
 		if err != nil {
 			return err
 		}
-	} else {
+	case noDiscovery:
+		logger.Info("discovery: disabled, keeping only trusted and inbound peers")
+		disc = discovery.Disabled{}
+	default:
```
Comment (momosh-ssv) [inline network/p2p/p2p_setup.go:312]: Any reason we're not erroring on an unknown value here? With three named modes, `default:` means `Discovery: nome` silently runs discv5.

The bigger one: a `none` node with no `TrustedPeers` starts, connects to nobody and still passes `Healthy()`, since `Disabled.Bootstrap` never sets `discoveryFailed` and there is no peer-count check anywhere. Erroring on unknown values, plus a Warn (not Info) when `none` has no trusted peers, would keep the mode honest.

On your note in the description: I'd keep `none` over a test-only hook on `Config`. It is a real setup for a trusted-peers-only node, and a named mode is easier to reason about than a field that exists only for tests.

File: network/p2p/testutils.go [hunk 3]
```diff
@@ -36,39 +36,32 @@ type LocalNet struct {
...
 		if err != nil {
 			return nil, err
 		}
 
+		for i, node := range ln.Nodes {
+			if err := node.Start(); err != nil {
+				return ln, fmt.Errorf("could not start node %d: %w", i, err)
+			}
+		}
+		if err := connectMesh(pCtx, ln.Nodes); err != nil {
```
Comment (momosh-ssv) [inline network/p2p/testutils.go:57]: Might be worth a timeout on this. The dials run on `pCtx`, which both callers build with `context.WithCancel(t.Context())`, so nothing in `connectMesh` sits under the 15s budget below. libp2p's `DialPeerTimeout` bounds each edge at 60s so it cannot hang forever, but a stalled dial burns that outside the retry loop, whose only escape is `pCtx.Done()`. The old code had no unbounded phase. `context.WithTimeout(pCtx, 15*time.Second)` would match the settle-wait budget.

File: network/p2p/testutils.go [hunk 4]
```diff
@@ -77,13 +70,10 @@ func CreateAndStartLocalNet(pCtx context.Context, logger *zap.Logger, options Lo
 					peers = node.(HostProvider).Host().Network().Peers()
 					select {
 					case <-ctx.Done():
-						return ctx.Err()
+						return fmt.Errorf("could not find enough peers for node %d, nodes quantity = %d, found = %d: %w", i, options.Nodes, len(peers), ctx.Err())
```
Comment (momosh-ssv) [inline network/p2p/testutils.go:73]: Could this report a failure for a node that already had enough peers? The loop condition is only re-checked after the select, so a read that satisfies `MinConnected` still falls into it, and if `ctx` is already done (a sibling failed and cancelled the errgroup) `ctx.Done()` wins over the 100ms timer. The node then reports `could not find enough peers ... found = 3` with `MinConnected = 1`, pointing the reader at the wrong node. `if len(peers) >= options.MinConnected { break }` right after the read covers it.

Worth noting the reworked message is otherwise a real improvement: `len(peers)` is read in the same iteration now, so the stale value the old post-loop block reported is gone.

Separately, now that the harness guarantees a full mesh, `MinConnected: n/2 - 1` in `p2p_test.go` is a weaker post-condition than what the code provides and would not catch a partially wired mesh. `nodes - 1`, as `p2p_validation_test.go` already uses, would.

File: network/discovery/disabled_test.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+	"github.com/stretchr/testify/require"
+)
+
+// Disabled discovers nothing: Bootstrap never hands out a peer, FindPeers yields a closed channel, and the
+// subnet and ENR hooks report nothing to update.
+func TestDisabled(t *testing.T) {
+	var d Service = Disabled{}
+
+	require.NoError(t, d.Bootstrap(func(PeerEvent) { t.Fatal("no peer is ever discovered") }))
```
Comment (momosh-ssv) [inline network/discovery/disabled_test.go:14]: Should we assert this differently? `Disabled.Bootstrap` ignores its parameter entirely, so this callback is unreachable by construction and would not catch a regression that did hand a peer to the handler. Counting calls and asserting `require.Zero` after `Bootstrap` returns puts the assertion on observed state instead.

File: network/p2p/config.go [hunk 6]
```diff
@@ -26,15 +26,19 @@ import (
...
+	noDiscovery    = "none"
 	minPeersBuffer = 10
 )
 
 // Config holds the configuration options for p2p network
 type Config struct {
 	Ctx          context.Context
 	Bootnodes    string   `yaml:"Bootnodes" env:"BOOTNODES" env-description:"Bootnodes to use for discovery (semicolon-separated ENRs, e.g. 'enr:-abc123;enr:-def456')" `
-	Discovery    string   `yaml:"Discovery" env:"P2P_DISCOVERY" env-description:"Discovery protocol to use (discv5, mdns)" `
+	Discovery    string   `yaml:"Discovery" env:"P2P_DISCOVERY" env-description:"Discovery protocol to use (discv5, mdns, or none to run without discovery)" `
```
Comment (momosh-ssv) [inline network/p2p/config.go:41]: Do we need to document `none` somewhere an operator would actually look? This tag is currently its only mention: `docs/DEV_GUIDE.md`, `cli/GENERATE_CONFIG.md` and `config/config.exporter.example.yaml` all still present `mdns` as the local-network option, and `cli/generate_config.go` still defaults to it. Given a `none` node is silently peerless without `TrustedPeers`, a line saying so would be worth adding.

File: network/p2p/testutils.go (hunk 1 above)
Comment (iurii-ssv) [reply network/p2p/testutils.go]: Set `DisableIPRateLimit: true` in `NewNetConfig` — it short-circuits `InterceptAccept`, so the mesh is no longer bounded by `ipLimitBurst` or the `MaxPeers`-coupled inbound limit, and it also zeroes the IP-colocation score penalty for the shared 127.0.0.1. Kept the balanced half-ring dialing (still dials each edge exactly once) and updated the comment.

Left the past-30s trimming degradation as-is: it's latent — tests settle and finish well before the trim timer — and addressing it means touching production trimming rather than the harness. Happy to follow up separately if we ever need long-running local nets.

File: network/p2p/p2p_setup.go (hunk 2 above)
Comment (iurii-ssv) [reply network/p2p/p2p_setup.go]: Both done in `setupDiscovery`: unknown modes now return an error instead of falling through to discv5 (empty string still maps to discv5, since `ApplyDefaults` seeds it before config load), and `none` with no `TrustedPeers` now logs a Warn. Agreed on keeping `none` as a real mode over a test-only hook.

File: network/p2p/testutils.go (hunk 3 above)
Comment (iurii-ssv) [reply network/p2p/testutils.go]: Done — `connectMesh` now runs under `context.WithTimeout(pCtx, 15s)`, matching the settle-wait budget, so a stalled dial can't burn the 60s `DialPeerTimeout` outside the retry loop.

File: network/p2p/testutils.go (hunk 4 above)
Comment (iurii-ssv) [reply network/p2p/testutils.go:83]: Fixed — the loop breaks on a satisfying read before the `select`, so a node that already has enough peers isn't reported as failed when a sibling cancels the errgroup ctx first. Also set `MinConnected` in `p2p_test.go` to `n-1` to assert the full mesh the harness now guarantees.

File: network/discovery/disabled_test.go (hunk 5 above)
Comment (iurii-ssv) [reply network/discovery/disabled_test.go]: Changed to assert on observed state — the handler increments a counter and the test requires it stayed zero after `Bootstrap` returns.

File: network/p2p/config.go (hunk 6 above)
Comment (iurii-ssv) [reply network/p2p/config.go:41]: Documented `none` in the operator-facing config docs — `config/config.exporter.example.yaml` and `cli/GENERATE_CONFIG.md` (the `--discovery` flag and the `Discovery` field) — including the peerless-without-`TrustedPeers` caveat. Left `docs/DEV_GUIDE.md` on `mdns`, since that guide is the local-cluster setup where mdns is the right choice.

File: network/p2p/testutils.go (hunk 1 above)
Comment (momosh-ssv) [reply network/p2p/testutils.go]: Verified the mesh fix: n=18, 24 and 32 all form a full mesh with `MinConnected: n-1`, and flipping `DisableIPRateLimit` back to `false` fails n=18 deterministically at the same edge I hit before (`connections per ip limit exceeded`, node 8 to node 9).

Might be worth correcting the deferral record though, because both of our diagnoses were off. I ran a subscribed 8-node local net and sampled peer counts:

```
t=0s    [7 7 7 7 7 7 7 7]
t=35s   [2 2 2 1 2 2 2 3]
t=70s   [0 0 0 0 0 0 0 0]
```

It does not degrade, it goes to zero and never recovers. And it is not `peersTrimming` or the `MaxPeers` coupling I pointed at: the run logs 32 lines of `disconnecting from peer with irrelevant subnets`. `DisconnectFromIrrelevantPeers` treats a peer whose subnets are unknown as irrelevant, and peer subnets are only fed by the discv5 ENR or the handshake. The handshake runs before the test subscribes, so every peer's subnets stay empty for the life of the net. mDNS hid this because rediscovery re-dialed and re-handshook after subscription. I also tested the harness-side fix I would have proposed (`MaxPeers = 2*options.Nodes`) and it does not help.

Still fine to defer for this PR, since the tests settle in ~8s. Worth folding the real mechanism into the deferral note or a TODO, so the next person adding a longer-running local-net test is not debugging a silent zero-peer mesh.

File: network/p2p/p2p_setup.go (hunk 2 above)
Comment (momosh-ssv) [reply network/p2p/p2p_setup.go]: Both confirmed in 9a2bba60. Two follow-ups, one small and one an outright nit.

Do we need a test on the new error path? `setupDiscovery`'s `default:` returns before it touches `Host()` or anything else on the struct, so a small table test over `discv5`, `mdns`, `none`, `""` and one junk value would lock the behavior down cheaply. Nothing covers it today.

Nit, flagging it as a nit and happy either way: `setupDiscovery` now reads `case discv5Discovery, ""`, so empty is explicitly discv5, but `configureAddrs` (`network/p2p/config.go:157`) still gates the external IP listener on `c.Discovery == discv5Discovery` exactly. An explicitly empty `Discovery` therefore runs discv5 without that listen addr. Practical impact looks like nothing, since libp2p expands the `0.0.0.0` listener across interfaces in `Addrs()` and discv5 advertises `discV5Opts.IP` separately, so this is consistency of intent rather than a bug.

**Review by momosh-ssv — DISMISSED** [review]

File: network/p2p/testutils.go (hunk 1 above)
Comment (iurii-ssv) [reply network/p2p/testutils.go]: Thanks for measuring it. I instrumented the same subscribed 8-node net, and the empty subnets turned out to be the node's own, not its peers': the handshake records every peer's subnets (`[11 64 90 114]` here), but only `UpdateSubnets` sets the node's current subnets, and the harness never ran it (the operator node does). So `DisconnectFromIrrelevantPeers` compared every peer against an empty set. Fixing only that brought back the `MaxPeers == Nodes` trimming you first flagged:

| variant | t=0 | t=35s | t=65s | t=95s |
|---|---|---|---|---|
| as-is | 7 | 0–2 | 0 | 0 |
| `UpdateSubnets` on every node | 7 | 1–3 | 1–3 | 1–3 |
| `MaxPeers = 2n` only | 7 | 0–2 | 0 | 0 |
| both | 7 | 7 | 7 | 7 |

The fix turned out to be harness-only, so I made it rather than deferring: every node now runs `UpdateSubnets`, and `MaxPeers` is `Nodes + maxPeersToDrop`, which keeps a full mesh clear of the trimming threshold for any node count (`2*Nodes` would still trim a 2-node net). `TestLocalNet_MeshSurvivesPeerTrimming` runs trimming rounds on a settled net and requires the full mesh to survive them; it fails with either half of the fix reverted.

File: network/p2p/p2p_setup.go (hunk 2 above)
Comment (iurii-ssv) [reply network/p2p/p2p_setup.go]: Both done. `TestSetupDiscovery` is a table over `discv5`, unset, `mdns`, `none` and `nome`: each valid mode must install its service, and `nome` must error without installing one. The discv5 rows build a real `DiscV5Service` on a free UDP port with the bootnodes stripped, so nothing leaves the host. Checked that it bites: switching on the raw field fails the unset row, and letting unknown modes fall through to discv5 fails the `nome` row.

On the nit: `setupDiscovery` and `configureAddrs` now both resolve the mode through `Config.discoveryMode()`, which maps unset to discv5, so an unset `Discovery` gets the explicit IP listener too and the two can't drift apart again.
